File: ruby/lib/trailer_vote/media_types/types/formatted_url.rb
# frozen_string_literal: true

module TrailerVote
  module MediaTypes
    module Types
      module_function

      class FormattedUrl

        class UnexpectedFormatVariable < ::MediaTypes::Scheme::ValidationError
          def initialize(variables:, actual:)
            self.variables = variables
            self.actual = actual
          end

          def message
            format(
              'Expected a formatted url only containing %<variables>s, instead got %<actual>s',
              variables: variables,
              actual: actual
            )
          end

          private

          attr_accessor :variables, :actual
        end

        def initialize(*variables)
          @variables = variables.map { |variable| "{#{variable}}" }
          variable_options = @variables.map { |variable| Regexp.escape(variable) }
          sections_first = "(?>#{variable_options.push('[^\s\/$.?#\{\}]').join('|')})"
          sections_not_first = "(?>#{variable_options.push('[^\\s\\{\\}]').join('|')})"

          @test = %r{^https?:\/\/#{sections_first}#{sections_not_first}*$}
        end

        def ===(other)
          return true if @test.match?(other)
          raise UnexpectedFormatVariable.new(variables: @variables, actual: other)
        end
      end

      # noinspection RubyInstanceMethodNamingConvention
      #
      def makeFormattedUrl(*variables) # rubocop:disable Naming/MethodName
        FormattedUrl.new(*variables)
      end
    end
  end
end

File: typescript/application/types/formatted_url.ts
// tslint:disable:max-classes-per-file
function escape(regex: string) {
  return regex.replace(/[-\/\\^$*+?.()|[\]{}]/g, '\\$&')
}

class UnexpectedFormatVariable extends Error {
  public variables: Readonly<string[]>
  public actual: Readonly<string>
  constructor({ variables, actual }: { variables: Readonly<string[]>, actual: Readonly<string> }) {
    super(`'Expected a formatted url only containing ${variables}, instead got ${actual}'`)

    this.variables = variables
    this.actual = actual
  }
}

export class FormattedUrl {
  private variables: string[]
  private test: RegExp

  constructor(...variables: string[]) {
    this.variables = variables.map((variable) => `{${variable}}`)
    const variableOptions = this.variables.map((variable) => escape(variable))
    const sectionsFirst = `(?>${variableOptions.concat(['[^\s\/$.?#\{\}]']).join('|')})`
    const sectionsNotFirst = `(?>${variableOptions.concat(['[^\\s\\{\\}]']).join('|')})`

    this.test = new RegExp(`^https?:\/\/${sectionsFirst}${sectionsNotFirst}*$/`)
  }

  public equals(other: string) {
    if (this.test.test(other)) {
      return true
    }

    throw new UnexpectedFormatVariable({ variables: this.variables, actual: other })
  }
}
